README.md
# CryptoLib (PHP)

[![Latest Version on Packagist](https://img.shields.io/packagist/v/ahsankhatri/cryptolib-php.svg?style=flat-square)](https://packagist.org/packages/ahsankhatri/cryptolib-php)
[![Total Downloads](https://img.shields.io/packagist/dt/ahsankhatri/cryptolib-php.svg?style=flat-square)](https://packagist.org/packages/ahsankhatri/cryptolib-php)
[![License](https://poser.pugx.org/ahsankhatri/cryptolib-php/license?format=flat-square)](https://packagist.org/packages/ahsankhatri/cryptolib-php)


This is the library for encrypting data with a key (password will be generate as per your parameters set) in PHP.

**WHY ANOTHER LIBRARY?** This was intended to developed for cross-platform AES Encryption [here](https://github.com/skavinvarnan/Cross-Platform-AES) as PHP was missing. My main objective is to create library for `AES-256-CBC` to contribute PHP package for Cross-Platform-AES package, more features will be added whenever I gets time.

### Features

- Support for Random IV (initialization vector) for encryption and decryption. Randomization is crucial for encryption schemes to achieve semantic security, a property whereby repeated usage of the scheme under the same key does not allow an attacker to infer relationships between segments of the encrypted message.
- Support for SHA-256 for hashing the key. Never use plain text as encryption key. Always hash the plain text key and then use for encryption. AES permits the use of 256-bit keys. Breaking a symmetric 256-bit key by brute force requires 2^128 times more computational power than a 128-bit key. A device that could check a billion billion (10^18) AES keys per second would in theory require about 3×10^51 years to exhaust the 256-bit key space.
- PHP-7 Support since `mcrypt` has been deprecated.

## Installation

You can install the package via composer:

```bash
composer require ahsankhatri/cryptolib-php
```

## Dependencies

The bindings require the following extensions in order to work properly:

- [`openssl`](https://secure.php.net/manual/en/book.openssl.php)

If you use Composer, these dependencies should be handled automatically. If you install manually, you'll want to make sure that these extensions are available.

## Usage

#### With Random IV

``` php
$string     = 'The quick brown fox jumps over to the lazy dog.';
$secretyKey = 'BlVssQKxzAHFAUNZbqvwS+yKw/m';

$encryption = new \MrShan0\CryptoLib\CryptoLib();

$cipher  = $encryption->encryptPlainTextWithRandomIV($string, $secretyKey);
echo 'Cipher: ' . $cipher . PHP_EOL;

$plainText = $encryption->decryptCipherTextWithRandomIV($cipher, $secretyKey);
echo 'Decrypted: ' . $plainText . PHP_EOL;
```

#### With Generated IV

``` php
$string     = 'The quick brown fox jumps over to the lazy dog.';
$secretyKey = 'BlVssQKxzAHFAUNZbqvwS+yKw/m';

$encryption = new \MrShan0\CryptoLib\CryptoLib();
$iv         = $encryption->generateRandomIV();

$cipher = $encryption->encrypt($string, $secretyKey, $iv);
echo 'Cipher: ' . $cipher . PHP_EOL;

$plainText = $encryption->decrypt($cipher, $secretyKey, $iv);
echo 'Decrypted: ' . $plainText . PHP_EOL;
```

### Testing

``` bash
composer test
```

### Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

### Security

If you discover any security related issues, please email ahsankhatri1992@gmail.com instead of using the issue tracker.

## Credits

- [Ahsaan Muhammad Yousuf](https://ahsaan.me)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `cryptolib-php` will be documented in this file

## 1.0.0 - 2018-12-05

- Initial release


tests/Encryption/AES-256-CBC.php
<?php

namespace MrShan0\CryptoLib\Tests\Encryption;

class AES_256_CBC_Test extends GenericTestCases
{
    protected $cryptolib;

    public function initClass()
    {
        $secretKey       = openssl_random_pseudo_bytes(16);
        $this->cryptolib = new \MrShan0\CryptoLib\CryptoLib([
            'method' => 'AES-256-CBC',
        ]);

        // Do not change this return order.
        return [
            $this->getCryptoLib(),
            $secretKey,
        ];
    }

    /**
     * @test
     *
     * @dataProvider additionProvider
     */
    public function random_iv($cryptolib, $key)
    {
        parent::random_iv($cryptolib, $key);
    }

    /**
     * @test
     *
     * @dataProvider additionProvider
     */
    public function huge_string_random_iv($cryptolib, $key)
    {
        parent::huge_string_random_iv($cryptolib, $key);
    }

    /**
     * @test
     *
     * @dataProvider additionProvider
     */
    public function custom_iv($cryptolib, $key)
    {
        parent::custom_iv($cryptolib, $key);
    }

    /**
     * @test
     *
     * @dataProvider additionProvider
     */
    public function huge_string_custom_iv($cryptolib, $key)
    {
        parent::huge_string_custom_iv($cryptolib, $key);
    }

    /**
     * @test
     *
     * @dataProvider additionProvider
     */
    public function diff_key_expect_exception($cryptolib, $key)
    {
        parent::diff_key_expect_exception($cryptolib, $key);
    }

    /**
     * @test
     *
     * @dataProvider additionProvider
     */
    public function huge_string_diff_key_expect_exception($cryptolib, $key)
    {
        parent::huge_string_diff_key_expect_exception($cryptolib, $key);
    }

}


tests/Encryption/GenericTestCases.php
<?php

namespace MrShan0\CryptoLib\Tests\Encryption;

use MrShan0\CryptoLib\Exceptions\UnableToDecrypt;
use PHPUnit\Framework\TestCase;

abstract class GenericTestCases extends TestCase
{
    abstract public function initClass();

    public function additionProvider()
    {
        return array(
            $this->initClass(),
        );
    }

    public function getCryptoLib()
    {
        return $this->cryptolib;
    }

    protected function random_iv($cryptolib, $key)
    {
        $plainText = openssl_random_pseudo_bytes(16);

        $cipher  = $cryptolib->encryptPlainTextWithRandomIV($plainText, $key);
        $decoded = $cryptolib->decryptCipherTextWithRandomIV($cipher, $key);

        $this->assertEquals($plainText, $decoded);
    }

    protected function huge_string_random_iv($cryptolib, $key)
    {
        $plainText = openssl_random_pseudo_bytes(16000);

        $cipher  = $cryptolib->encryptPlainTextWithRandomIV($plainText, $key);
        $decoded = $cryptolib->decryptCipherTextWithRandomIV($cipher, $key);

        $this->assertEquals($plainText, $decoded);
    }

    protected function custom_iv($cryptolib, $key)
    {
        $plainText = openssl_random_pseudo_bytes(16);
        $iv               = $cryptolib->generateRandomIV();

        $cipher  = $cryptolib->encrypt($plainText, $key, $iv);
        $decoded = $cryptolib->decrypt($cipher, $key, $iv);

        $this->assertEquals($plainText, $decoded);
    }

    protected function huge_string_custom_iv($cryptolib, $key)
    {
        $plainText = openssl_random_pseudo_bytes(16000);
        $iv               = $cryptolib->generateRandomIV();

        $cipher  = $cryptolib->encryptPlainTextWithRandomIV($plainText, $key, $iv);
        $decoded = $cryptolib->decryptCipherTextWithRandomIV($cipher, $key, $iv);

        $this->assertEquals($plainText, $decoded);
    }

    protected function diff_key_expect_exception($cryptolib, $key)
    {
        $this->expectException(UnableToDecrypt::class);

        $plainText = openssl_random_pseudo_bytes(16);

        $cipher  = $cryptolib->encrypt($plainText, openssl_random_pseudo_bytes(16), $cryptolib->generateRandomIV());
        $decoded = $cryptolib->decrypt($cipher, openssl_random_pseudo_bytes(16), $cryptolib->generateRandomIV());
    }

    protected function huge_string_diff_key_expect_exception($cryptolib, $key)
    {
        $this->expectException(UnableToDecrypt::class);

        $plainText = openssl_random_pseudo_bytes(16000);

        $cipher  = $cryptolib->encrypt($plainText, openssl_random_pseudo_bytes(16), $cryptolib->generateRandomIV());
        $decoded = $cryptolib->decrypt($cipher, openssl_random_pseudo_bytes(16), $cryptolib->generateRandomIV());
    }

}


src/CryptoLib.php
<?php

namespace MrShan0\CryptoLib;

use MrShan0\CryptoLib\Exceptions\NotSecureIVGenerated;
use MrShan0\CryptoLib\Exceptions\UnableToDecrypt;

class CryptoLib
{
    /**
     * Default setting for generating hash
     *
     * @var array
     */
    private $_opts = [

        /**
         * Crypto method to encrypt your string with
         */
        'method'             => 'AES-256-CBC',

        /**
         * Algorithm to apply while generating password via secrety key
         */
        'algorithm'          => 'sha256',

        /**
         * Maximum key size
         */
        'maxKeySize'         => 32,

        /**
         * Maximum IV size
         */
        'maxIVSize'          => 16,

        /**
         * Number of iterations to generate hash for password
         */
        'numberOfIterations' => 1,
    ];

    public function __construct(array $options = array())
    {
        $this->_opts = $options + $this->_opts;
    }

    public function getOptions($key = '')
    {
        if ($key) {
            return $this->_opts[$key];
        }

        return $this->_opts;
    }

    public function generateRandomIV()
    {
        $efforts     = 0;
        $maxEfforts  = 50;
        $wasItSecure = false;

        do {

            $efforts += 1;
            $iv = openssl_random_pseudo_bytes(openssl_cipher_iv_length($this->getOptions('method')), $wasItSecure);
            if ($efforts == $maxEfforts) {
                throw new NotSecureIVGenerated('Unable to genereate secure IV.');
                break;
            }

        } while (!$wasItSecure);

        return $iv;
    }

    public function getComputedHash($key)
    {
        $hash = $key;
        for ($i = 0; $i < intval($this->getOptions('numberOfIterations')); $i++) {
            $hash = hash($this->getOptions('algorithm'), $hash);
        }

        return $hash;
    }

    public function encrypt($plainText, $key, $initVector)
    {
        return $this->encryptOrDecrypt('encrypt', $plainText, $key, $initVector);
    }

    public function decrypt($encryptedText, $key, $initVector)
    {
        $plainText = $this->encryptOrDecrypt('decrypt', $encryptedText, $key, $initVector);

        if (false === $plainText) {
            throw new UnableToDecrypt('Unable to decrypt your encrypted string.');
        }

        return $plainText;
    }

    public function encryptPlainTextWithRandomIV($plainText, $key)
    {
        return $this->encryptOrDecrypt('encrypt', $this->generateRandomIV() . $plainText, $key, $this->generateRandomIV());
    }

    public function decryptCipherTextWithRandomIV($cipherText, $key)
    {
        return substr(
            $this->encryptOrDecrypt('decrypt', $cipherText, $key, $this->generateRandomIV()),
            intval($this->getOptions('maxIVSize'))
        );
    }

    private function encryptOrDecrypt($mode, $string, $key, $initVector)
    {
        $password = substr($this->getComputedHash($key), 0, intval($this->getOptions('maxKeySize')));

        if ('encrypt' === $mode) {
            return base64_encode(openssl_encrypt(
                $string,
                $this->getOptions('method'),
                $password,
                OPENSSL_RAW_DATA,
                $initVector
            ));
        }

        return openssl_decrypt(
            base64_decode($string),
            $this->getOptions('method'),
            $password,
            OPENSSL_RAW_DATA,
            $initVector
        );
    }

}
